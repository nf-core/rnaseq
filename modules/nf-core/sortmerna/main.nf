process SORTMERNA {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/sortmerna:4.3.6--h9ee0642_0' :
        'biocontainers/sortmerna:4.3.6--h9ee0642_0' }"

    input:
    tuple val(meta), path(reads)
    path  fastas

    output:
    tuple val(meta), path("*non_rRNA.fastq.gz"), emit: reads
    tuple val(meta), path("*.log")     , emit: log
    path  "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args        = task.ext.args  ?: ''
    def prefix       = task.ext.prefix ?: "${meta.id}"
    def reads_input = reads instanceof List ? reads.collect{"--reads $it"}.join(' ') : "--reads $reads"
    def n_fastq     = reads instanceof List ? reads.size() : 1
    if ( n_fastq == 1 ) {
        mv_cmd = "mv non_rRNA_reads.f*q.gz ${prefix}.non_rRNA.fastq.gz"
        paired_cmd = ''
        out2_cmd   = ''
    } else {
        mv_cmd = """
        mv non_rRNA_reads_fwd.f*q.gz ${prefix}_1.non_rRNA.fastq.gz
        mv non_rRNA_reads_rev.f*q.gz ${prefix}_2.non_rRNA.fastq.gz
        """.stripIndent()
        paired_cmd = "--paired_in"
        out2_cmd   = "--out2"
    }
    """
    sortmerna \\
        ${'--ref '+fastas.join(' --ref ')} \\
        $reads_input \\
        --threads $task.cpus \\
        --workdir . \\
        --aligned rRNA_reads \\
        --fastx \\
        --other non_rRNA_reads \\
        $paired_cmd \\
        $out2_cmd \\
        $args

    $mv_cmd
    mv rRNA_reads.log ${prefix}.sortmerna.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sortmerna: \$(echo \$(sortmerna --version 2>&1) | sed 's/^.*SortMeRNA version //; s/ Build Date.*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def reads_input = reads instanceof List ? reads.collect{"--reads $it"}.join(' ') : "--reads $reads"
    def n_fastq     = reads instanceof List ? reads.size() : 1
    if ( n_fastq == 1 ) {
        mv_cmd = "touch ${prefix}.non_rRNA.fastq.gz"
    } else {
        mv_cmd = """
        touch ${prefix}_1.non_rRNA.fastq.gz
        touch ${prefix}_2.non_rRNA.fastq.gz
        """.stripIndent()
    }
    """
    $mv_cmd
    touch ${prefix}.sortmerna.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sortmerna: \$(echo \$(sortmerna --version 2>&1) | sed 's/^.*SortMeRNA version //; s/ Build Date.*\$//')
    END_VERSIONS
    """
}
