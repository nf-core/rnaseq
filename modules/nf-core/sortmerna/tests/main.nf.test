nextflow_process {

    name "Test Process SORTMERNA"
    script "../main.nf"
    process "SORTMERNA"
    tag "modules"
    tag "modules_nfcore"
    tag "sortmerna"

    test("sarscov2 single_end") {

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true) ]
                ]
                input[1] = [ file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.fasta", checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.reads },
                { assert file(process.out.log[0][1]).text.contains("Total reads passing E-value threshold = 100 (100.00)") },
                {
                    assert snapshot(
                        (
                            [process.out.reads[0][0].toString()] + // meta
                            process.out.reads.collect { file(it[1]).getName() } +
                            process.out.log.collect { file(it[1]).getName() }
                        ).sort()
                    ).match("sarscov2 single_end_match")
                },
                { assert snapshot(process.out.versions).match("versions_single_end") }
            )
        }

    }

    test("sarscov2 single_end stub") {

        options '-stub'

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true) ]
                ]
                input[1] = [ file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.fasta", checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        (
                            [process.out.reads[0][0].toString()] + // meta
                            process.out.reads.collect { file(it[1]).getName() } +
                            process.out.log.collect { file(it[1]).getName() }
                        ).sort()
                    ).match("sarscov2 single_end-for_stub_match")
                },
                { assert snapshot(process.out.versions).match("versions_single_end_stub") }
            )
        }

    }

    test("sarscov2 paired_end") {

        when {
            process {
                """
                input[0] = [ 
                    [ id:'test' ], // meta map
                    [ 
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true),
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [ file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.fasta", checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.reads },
                { assert file(process.out.log[0][1]).text.contains("Total reads passing E-value threshold = 200 (100.00)") },
                {
                    assert snapshot(
                        (
                            [process.out.reads[0][0].toString()] + // meta
                            process.out.reads.collect { it[1].collect { item -> file(item).getName() } } +
                            process.out.log.collect { file(it[1]).getName() }
                        ).sort()
                    ).match("sarscov2 paired_end_match")
                },
                { assert snapshot(process.out.versions).match("versions_paired_end") }
            )
        }

    }

    test("sarscov2 paired_end stub") {

        options '-stub'

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                    [ 
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true),
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [ file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.fasta", checkIfExists: true) ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        (
                            [process.out.reads[0][0].toString()] + // meta
                            process.out.reads.collect { it[1].collect { item -> file(item).getName() } } +
                            process.out.log.collect { file(it[1]).getName() }
                        ).sort()
                    ).match("sarscov2 paired_end-for_stub_match")
                },
                { assert snapshot(process.out.versions).match("versions_paired_end_stub") }
            )
        }

    }

}
