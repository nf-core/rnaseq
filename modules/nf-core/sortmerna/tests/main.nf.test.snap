{
    "sarscov2 single_end-for_stub_match": {
        "content": [
            [
                "test.non_rRNA.fastq.gz",
                "test.sortmerna.log",
                "{id=test}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:28:29.197913"
    },
    "sarscov2 paired_end_match": {
        "content": [
            [
                "{id=test}",
                [
                    "test_1.non_rRNA.fastq.gz",
                    "test_2.non_rRNA.fastq.gz"
                ],
                "test.sortmerna.log"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:28:49.914992"
    },
    "versions_paired_end_stub": {
        "content": [
            [
                "versions.yml:md5,5a099f3eea59b9deee96deffcf3387f5"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:27:11.223149"
    },
    "versions_paired_end": {
        "content": [
            [
                "versions.yml:md5,5a099f3eea59b9deee96deffcf3387f5"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:27:04.517155"
    },
    "versions_single_end": {
        "content": [
            [
                "versions.yml:md5,7df9d50209f351e1f75e05a1fad6ba4b"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:10:22.534568982"
    },
    "versions_single_end_stub": {
        "content": [
            [
                "versions.yml:md5,7df9d50209f351e1f75e05a1fad6ba4b"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:10:35.228450189"
    },
    "sarscov2 single_end_match": {
        "content": [
            [
                "test.non_rRNA.fastq.gz",
                "test.sortmerna.log",
                "{id=test}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:28:23.20327"
    },
    "sarscov2 paired_end-for_stub_match": {
        "content": [
            [
                "{id=test}",
                [
                    "test_1.non_rRNA.fastq.gz",
                    "test_2.non_rRNA.fastq.gz"
                ],
                "test.sortmerna.log"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-07T12:28:56.063579"
    }
}